app/models/user_snp.rb

```diff
@@ -1,9 +1,37 @@
-class UserSnp < ActiveRecord::Base
-  self.primary_keys = [:genotype_id, :snp_name]
-  belongs_to :snp, foreign_key: :snp_name, primary_key: :name, counter_cache: true
-  has_one :user, through: :genotype
-  belongs_to :genotype
-
-  validates_presence_of :snp
-  validates_presence_of :genotype
+class UserSnp
+  attr_reader :snp, :genotype
+  delegate :user, to: :genotype
+
+  def initialize(snp, genotype, local_genotype = nil)
+    @snp = snp
+    @genotype = genotype
+    @local_genotype = local_genotype if local_genotype
+  end
+
+  def local_genotype
+    return @local_genotype if defined?(@local_genotype)
+
+    snp_name = ActiveRecord::Base.sanitize(snp.name)
+    @local_genotype = Genotype.unscoped
+                              .select("snps -> #{snp_name} AS local_genotype")
+                              .where(id: genotype.id)
```

Comment:
Use 2 (not 26) spaces for indenting an expression in an assignment spanning multiple lines.